proxy/__init__.py

```diff
@@ -0,0 +1,201 @@
+"""
+The proxy module provides the means to proxy images from other registry instances.
+Registries following the distribution spec are supported.
+"""
+from __future__ import annotations
+from typing import Any
+import re
+import json
+
+import requests
+
+from app import model_cache
+from data.cache import cache_key
+from data.database import ProxyCacheConfig
+from digest.digest_tools import sha256_digest
+
+
+WWW_AUTHENTICATE_REGEX = re.compile(r'(\w+)[=] ?"?([^",]+)"?')
+TOKEN_VALIDITY_LIFETIME_S = 60 * 60  # 1 hour, in seconds - Quay's default
+TOKEN_RENEWAL_THRESHOLD = 10  # interval (in seconds) when to renew auth token
+
+
+REGISTRY_URLS = {"docker.io": "registry-1.docker.io"}
+
+
+class UpstreamRegistryError(Exception):
+    def __init__(self, detail):
+        msg = f"Error requesting upstream registry: {detail}"
+        super().__init__(msg)
+
+
+def parse_www_auth(value: str) -> dict[str, str]:
+    """
+    Parses WWW-Authenticate parameters and returns a dict of key=val.
+    See https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate
+    for details.
+    This parser is *not* fully compliant with RFC 7235, notably it does not support
+    multiple challenges.
+    """
+    scheme = value.split(" ", 1)[0]
+    matches = WWW_AUTHENTICATE_REGEX.findall(value)
+    parsed = dict(matches)
+    if scheme:
+        parsed["scheme"] = scheme
+    return parsed
+
+
+class Proxy:
+    def __init__(self, config: ProxyCacheConfig, repository: str):
+        self._config = config
+
+        hostname = REGISTRY_URLS.get(
+            config.upstream_registry_hostname,
+            config.upstream_registry_hostname,
+        )
+        url = f"https://{hostname}"
+        if config.insecure:
+            url = f"http://{hostname}"
+
+        self.base_url = url
+        self._session = requests.Session()
+        self._repo = repository
+        self._authorize(self._credentials())
+
+    def get_manifest(self, image_ref: str, media_type: str | None = None) -> tuple[str, str | None]:
+        headers = {}
+        if media_type is not None:
+            headers["Accept"] = media_type
+
+        url = f"{self.base_url}/v2/{self._repo}/manifests/{image_ref}"
+        resp = self.get(url, headers=headers)
+        raw_manifest = resp.content
+        content_type = resp.headers.get("content-type")
+        return raw_manifest, content_type
+
+    def manifest_exists(self, image_ref: str, media_type: str | None = None) -> str:
+        url = f"{self.base_url}/v2/{self._repo}/manifests/{image_ref}"
+        headers = {}
+        if media_type is not None:
+            headers["Accept"] = media_type
+        resp = self.head(url, headers=headers, allow_redirects=True)
+        return resp.headers.get("docker-content-digest")
```

Comment:
from OCI spec:
> the client MAY encounter Docker-specific headers, such as Docker-Content-Digest, or Docker-Distribution-API-Version. These headers are OPTIONAL and clients SHOULD NOT depend on them.
> — https://github.com/opencontainers/distribution-spec/blob/main/spec.md#definitions

Reply:
Yup, that's why I used `get` to retrieve it. I've changed the return type to `str | None` to reflect that (will push soon)

Reply:
The caller of this function shouldn't rely on this value, as `clients SHOULD NOT depend on` it. So it's safer to always return None.

Reply:
You make a good point.
The thing is that we're using this digest to check whether the manifest cached in quay is up-to-date with the upstream. If the upstream doesn't return a digest then we will end up pulling the manifest again.
I'm not sure what's the best way to do this. We could do a GET instead of a HEAD to compare digests, but on docker hub a GET will count towards rate limit while a HEAD won't.
Another option would be to do a GET only if we find that the `docker-content-digest` header from the HEAD came up empty, then calculate the manifest digest from the actual manifest. I'd argue that this is a nice-to-have.
Thoughts?

Reply:
Oh, I didn't realize `manifest_exists` returns the current digest of the tag. It might be worthy to comment this function.

Reply:
Good point, added comment.